src/Symfony/Bundle/FrameworkBundle/Routing/RouteCollectionBuilder.php

```diff
@@ -0,0 +1,433 @@
+     *
+     * @param string $controllerClass
+     *
+     * @return $this
+     */
+    public function setControllerClass($controllerClass)
+    {
+        if (!class_exists($controllerClass)) {
+            throw new \InvalidArgumentException(sprintf('The controller class "%s" does not exist.', $controllerClass));
+        }
+
+        $this->controllerClass = $controllerClass;
+
+        return $this;
+    }
+
+    /**
+     * Creates the final ArrayCollection, returns it, and clears everything.
+     *
+     * @return RouteCollection
+     */
+    public function build()
+    {
+        $routeCollection = new RouteCollection();
+
+        foreach ($this->routes as $name => $route) {
+            if ($route instanceof Route) {
+                // auto-generate the route name if it's been marked
+                if ('_unnamed_route_' === substr($name, 0, 15)) {
+                    $name = $this->generateRouteName($route);
+                }
+
+                $this->ensureRouteController($route);
+
+                $route->setDefaults(array_merge($this->defaults, $route->getDefaults()));
+                $route->setRequirements(array_merge($this->requirements, $route->getRequirements()));
+                $route->setOptions(array_merge($this->options, $route->getOptions()));
+
+                if (null !== $this->prefix) {
+                    $route->setPath('/'.$this->prefix.$route->getPath());
+                }
+
+                if (!$route->getHost()) {
+                    $route->setHost($this->host);
+                }
+
+                if (!$route->getCondition()) {
+                    $route->setCondition($this->condition);
+                }
+
+                if (!$route->getSchemes()) {
+                    $route->setSchemes($this->schemes);
+                }
+
+                if (!$route->getMethods()) {
+                    $route->setMethods($this->methods);
+                }
+
+                $routeCollection->add($name, $route);
+            } else {
+                /* @var self $route */
+
+                $subCollection = $route->build();
+                $subCollection->addPrefix($this->prefix);
+
+                $routeCollection->addCollection($subCollection);
+            }
+        }
+
+        foreach ($this->resources as $resource) {
+            $routeCollection->addResource($resource);
+        }
+
+        return $routeCollection;
+    }
+
+    /**
+     * Attempts to safely prefix controllers with the controller class if necessary.
+     *
+     * @param Route $route
+     */
+    private function ensureRouteController(Route $route)
+    {
+        // only do work if there is a controller class set
+        if (!$this->controllerClass) {
+            return;
+        }
+
+        $controller = $route->getDefault('_controller');
+
+        // only apply controller class to a (non-empty) string
+        if (!is_string($controller) || !$controller) {
```

Comment:
It's not quite the same thing - I'm also filtering for other things, like array()s